module/zfs/qat_compress.c

```diff
@@ -0,0 +1,381 @@
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+#ifdef HAVE_QAT
+#include <linux/slab.h>
+#include <linux/vmalloc.h>
+#include <linux/pagemap.h>
+#include <linux/completion.h>
+#include "qat_compress.h"
+
+#define	TIMEOUT_MS	500 /* 0.5 seconds */
+#define	INST_NUM	6
+#define	GZIP_HEAD_SZ	2
+#define	GZIP_FOOT_SZ	4
+
+static CpaInstanceHandle dc_inst_handles[INST_NUM];
+static CpaDcSessionHandle session_handles[INST_NUM];
+static Cpa16U num_inst = 0;
+static Cpa16U inst = 0;
+int qat_init_done = 0;
+
+#define	PHYS_CONTIG_ALLOC(pp_mem_addr, size_bytes)	\
+	mem_alloc_contig((void *)(pp_mem_addr), (size_bytes))
+
+#define	PHYS_CONTIG_FREE(p_mem_addr)	\
+	mem_free_contig((void *)&(p_mem_addr))
+
+static inline struct page *mem_to_page(void *addr)
+{
+	if (!is_vmalloc_addr(addr))
+		return (virt_to_page(addr));
+
+	return (vmalloc_to_page(addr));
+}
+
+static void qat_dc_callback(void *p_callback, CpaStatus status)
+{
+	if (NULL != p_callback)
+		complete((struct completion *)p_callback);
+}
+
+static __inline CpaStatus
+mem_alloc_contig(void **pp_mem_addr, Cpa32U size_bytes)
+{
+	*pp_mem_addr = kmalloc_node(size_bytes,
+		GFP_KERNEL, 0);
+	if (NULL == *pp_mem_addr)
+		return (CPA_STATUS_RESOURCE);
+	return (CPA_STATUS_SUCCESS);
+}
+
+static __inline void mem_free_contig(void **pp_mem_addr)
+{
+	if (NULL != *pp_mem_addr) {
+		kfree(*pp_mem_addr);
+		*pp_mem_addr = NULL;
+	}
+}
+
+int
+qat_init(void)
+{
+	CpaStatus status = CPA_STATUS_SUCCESS;
+	Cpa32U sess_size = 0;
+	Cpa32U ctx_size = 0;
+	Cpa32U buff_meta_size = 0;
+	CpaDcSessionSetupData sd = {0};
+	int i;
+
+	if (qat_init_done != 0)
+		return (0);
+
+	status = cpaDcGetNumInstances(&num_inst);
+	if (status != CPA_STATUS_SUCCESS || num_inst == 0)
+		return (-1);
+	status = cpaDcGetInstances(num_inst, &dc_inst_handles[0]);
+	if (status != CPA_STATUS_SUCCESS)
+		return (-1);
+
+	for (i = 0; i < num_inst; i++) {
+		cpaDcSetAddressTranslation(dc_inst_handles[i],
+		    (void*)virt_to_phys);
+
+		status = cpaDcBufferListGetMetaSize(dc_inst_handles[i], 1,
+		    &buff_meta_size);
+
+		status = cpaDcStartInstance(dc_inst_handles[i], 0, NULL);
+		if (status != CPA_STATUS_SUCCESS)
+			goto fail;
+
+		sd.compLevel = CPA_DC_L1;
+		sd.compType = CPA_DC_DEFLATE;
+		sd.huffType = CPA_DC_HT_STATIC;
+		sd.sessDirection = CPA_DC_DIR_COMBINED;
+		sd.sessState = CPA_DC_STATELESS;
+		sd.deflateWindowSize = 7;
+		sd.checksum = CPA_DC_ADLER32;
+		status = cpaDcGetSessionSize(dc_inst_handles[i],
+		    &sd, &sess_size, &ctx_size);
+		if (status != CPA_STATUS_SUCCESS)
+			goto fail;
+
+		PHYS_CONTIG_ALLOC(&session_handles[i], sess_size);
+		if (session_handles[i] == NULL)
+			goto fail;
+
+		status = cpaDcInitSession(dc_inst_handles[i],
+		    session_handles[i],
+		    &sd, NULL, qat_dc_callback);
+		if (status != CPA_STATUS_SUCCESS)
+			goto fail;
+	}
+
+	qat_init_done = 1;
+	printk(KERN_INFO "qat accel init success\n");
```

Comment:
Removed the printk.
For the statistics of the QAT device, the QAT driver has already implemented it, we can cat /proc/icp_dh895xxcc_dev/qat to get the QAT status. I've added one note in the kernel.m4 in the QAT config session. Will also add this usage in the commit log after rebase.